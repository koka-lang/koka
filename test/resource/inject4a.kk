// Test inject over handled effects

effect heap<a> {
  fun get() : a 
  fun set(value : a) : ()
}

val new-ref = handler(s) {
  get() -> resume(s,s)
  set(x) -> resume((),x)
}

fun test() {
  new-ref("hi") {
    new-ref(44) {
      // heap<int>,heap<string>|e>
      get() - inject<heap<int>>{(get():string).count}  // 44 - "hi".count
    }
  }
}

fun main() {
  println(test())
}