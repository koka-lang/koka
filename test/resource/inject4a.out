[01;30moperator branch (new-ref) get: resume tail
[00;39;49m[00m[01;30moperator branch (new-ref) set: resume tail
[00;39;49m[00m42
.Op-get : forall<a> .op-get<a>
.Op-set : forall<a> (value : a) -> .op-set<a>
.Ops-get : forall<a> (get : .op-get<a>) -> .ops-heap<a>
.Ops-set : forall<a> (set : .op-set<a>) -> .ops-heap<a>
.tag-heap : string
get : forall<a> () -> (heap<a>) a
main : () -> console ()
new-ref : forall<a,b,e> (s : a, action : () -> <heap<a>|e> b) -> e b
set : forall<a> (value : a) -> (heap<a>) ()
test : () -> int